### ISSUE TITLE
[windows-latest] CMake cannot find OpenSSL from env

### ISSUE THREAD

**Issue by Ace-Radom** [actions/runner-images#9395]
### Description

My project uses CMake and it needs OpenSSL to build. Therefore I have the following codes in my CMakeLists:

```cmake
find_package(OpenSSL REQUIRED)
if(NOT OPENSSL_FOUND)
    message(FATAL "Cannot find OpenSSL, please check environment")
endif()
```

This works on my local dev environment, but on GitHub-hosted runners it reports:

```
CMake Error at C:/Program Files/CMake/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:2[30](https://github.com/Ace-Radom/figlet4win/actions/runs/8016050930/job/21897382523#step:4:31) (message):
  Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the
  system variable OPENSSL_ROOT_DIR (missing: OPENSSL_CRYPTO_LIBRARY
  OPENSSL_INCLUDE_DIR)
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.28/Modules/FindOpenSSL.cmake:668 (find_package_handle_standard_args)
  CMakeLists.txt:[33](https://github.com/Ace-Radom/figlet4win/actions/runs/8016050930/job/21897382523#step:4:34) (find_package)
```

I checked my former actions and found out that this error didn't exist three days ago, and I supposed that GitHub-hosted windows server 2022 runners do have an OpenSSL installed. I'm wondering why it is happened.

### Platforms affected

- [ ] Azure DevOps
- [X] GitHub Actions - Standard Runners
- [ ] GitHub Actions - Larger Runners

### Runner images affected

- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [ ] macOS 11
- [ ] macOS 12
- [ ] macOS 13
- [ ] macOS 13 Arm64
- [ ] macOS 14
- [ ] macOS 14 Arm64
- [ ] Windows Server 2019
- [X] Windows Server 2022

### Image version and build link

Failed action links:

- https://github.com/Ace-Radom/figlet4win/actions/runs/8016050930/job/21897256407 (Version 20240218.2.0)
- https://github.com/Ace-Radom/figlet4win/actions/runs/8016050930/job/21897336520 (Version 20240218.2.0)
- https://github.com/Ace-Radom/figlet4win/actions/runs/8016050930/job/21897382523 (Version 20240218.2.0)

### Is it regression?

https://github.com/Ace-Radom/figlet4win/actions/runs/7957935446/job/21721752104 (Version 20240211.1.0)

### Expected behavior

CMake should find OpenSSL automatically, without `OPENSSL_ROOT_DIR` set. It seems that this bug only exists in the latest image version (20240218.2.0).

### Actual behavior

CMake cannot find OpenSSL from the latest windows server 2022 runner. Build fails because OpenSSL is required.

### Repro steps

1. Find OpenSSL in CMakeLists and set it to `REQUIRED`
2. CMake cannot find it, FATAL occurs and exits

**Comment by sergei-pyshnoi** [comment]
Hello @Ace-Radom . We will take a look

**Comment by mikhailkoliada** [comment]
Hey! Looks like cake 3.28 update got some changes connected with how it treats mingw generator (cmake is the only component which got updated in the `mingw-openssl-cmake` chain since the previous runner image release.

<img width="653" alt="image" src="https://github.com/actions/runner-images/assets/88318005/96142ab9-4479-4a5c-9659-cab4d8810649">

Completing your above snippet to the complete project:

```csl shell=bash
project(ossl)
cmake_minimum_required(VERSION 3.0)
find_package(OpenSSL REQUIRED)
if(NOT OPENSSL_FOUND)
    message(FATAL "Cannot find OpenSSL, please check environment")
endif()
```

Then running the build:

```
 runs-on: windows-latest
    steps:
      - uses: actions/checkout@v4
      - name: test
        run: |
         Get-ChildItem
         choco uninstall cmake.install -x
         choco install cmake.install --version=3.27.9
         cmake -Wno-dev -G "MinGW Makefiles" .

```

fixes the problem, interestingly, the problem is only reproducible with the MinGw makefile generator, if the standard Visual Studio one is used, openssl is found during runtime flawlessly. I do not believe we can do anything to fix make sources from our side so please report the bug [cmake upstream](https://gitlab.kitware.com/cmake/cmake/-/issues).

Feel free to reach us again if you have questions left.

**Comment by Ace-Radom** [comment]
Yes this problem is solved. I'll do some further tests on the problem with CMake 3.28 you mentioned and report them to the CMake dev team maybe later.
Anyway thanks for your quick and detailed response xD.

**Comment by Ace-Radom** [comment]
Just FYI: I reported this problem to CMake dev team and got this answer from them.

> Thanks.  Being in the PATH is enough for the command-line tools to be found, but in general it is not enough for the associated headers and libraries to be found by development tools.  CMake incorrectly used PATH for the latter on Windows anyway, which caused many problems as described in [#24216 (closed)](https://gitlab.kitware.com/cmake/cmake/-/issues/24216), so [!8802 (merged)](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/8802) removed the behavior.  Unfortunately there is no good way to provide compatibility or a transition period while also satisfying the reasons for the change.  The change is called out in the 3.28 release notes.
If your project depends on OpenSSL's headers and libraries, it is up to you to make sure they are available as a dependency, with appropriate environment variables set so that development tools can find them.  If the GitHub Actions environment you're using intends for OpenSSL's development components to be available for use by builds, it needs to set additional environment variables to tell build systems where to find them.  If it intends only for OpenSSL's command-line tools to be available in the PATH for use by scripts, but happens to use an installation that comes with the headers and libraries too, then those are incidental and should not be relied upon as build dependencies.

That means it's not a bug but an intentional behavior change. It is also described in the [CMake 3.28 release note](https://cmake.org/cmake/help/v3.28/release/3.28.html#other-changes) where I found out later yesterday.
